backpack/core/derivatives/crossentropyloss.py

```diff
@@ -64,49 +73,73 @@ def _sqrt_hessian_sampled(
         sqrt_mc_h = (probs_unsqueezed - classes) / sqrt(M)
 
         if module.reduction == "mean":
-            N = module.input0.shape[0]
-            sqrt_mc_h /= sqrt(N)
+            sqrt_mc_h /= sqrt(self._get_mean_normalization(module.input0))
 
+        sqrt_mc_h = self._ungroup_batch_and_additional(sqrt_mc_h, *rearrange_info)
         return sqrt_mc_h
 
-    def _sum_hessian(self, module, g_inp, g_out):
+    def _sum_hessian(
+        self, module: CrossEntropyLoss, g_inp: Tuple[Tensor], g_out: Tuple[Tensor]
+    ) -> Tensor:
         self._check_2nd_order_parameters(module)
 
         probs = self._get_probs(module)
-        sum_H = diag(probs.sum(0)) - einsum("bi,bj->ij", (probs, probs))
+
+        if probs.dim() == 2:
+            diagonal = diag(probs.sum(0))
+            sum_H = diagonal - einsum("nc,nd->cd", probs, probs)
+        else:
+            diagonal = diag(probs.sum(0).flatten()).reshape(
+                *probs.shape[1:], *probs.shape[1:]
+            )
+            sum_H = (
+                diagonal
+                - einsum(
+                    "ncx,ndy,xy->cxdy",
+                    probs.flatten(2),
+                    probs.flatten(2),
+                    eye(
+                        prod(probs.shape[2:]),
```

Comment:
I think you can do `probs.shape[2:].numel()` and get rid of the `prod` import

Reply:
Yes, that works. Thank you for this solution.